src/agent/src/image.rs

```diff
@@ -0,0 +1,382 @@
+}
+
+impl ImageService {
+    pub fn new() -> Self {
+        env::set_var("KATA_IMAGE_WORK_DIR", KATA_IMAGE_WORK_DIR);
+
+        Self {
+            image_client: Arc::new(Mutex::new(ImageClient::default())),
+            images: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    /// Get the singleton instance of image service.
+    pub async fn singleton() -> Result<ImageService> {
+        IMAGE_SERVICE
+            .lock()
+            .await
+            .clone()
+            .ok_or_else(|| anyhow!("image service is uninitialized"))
+    }
+
+    async fn add_image(&self, image: String, cid: String) {
+        self.images.lock().await.insert(image, cid);
+    }
+
+    /// Set proxy environment from AGENT_CONFIG
+    fn set_proxy_env_vars() {
+        let https_proxy = &AGENT_CONFIG.https_proxy;
+        if !https_proxy.is_empty() {
+            env::set_var("HTTPS_PROXY", https_proxy);
+        }
+        let no_proxy = &AGENT_CONFIG.no_proxy;
+        if !no_proxy.is_empty() {
+            env::set_var("NO_PROXY", no_proxy);
+        }
+    }
+
+    /// pause image is packaged in rootfs
+    fn unpack_pause_image(cid: &str, target_subpath: &str) -> Result<String> {
+        let cc_pause_bundle = Path::new(KATA_PAUSE_BUNDLE);
+        if !cc_pause_bundle.exists() {
+            return Err(anyhow!("Pause image not present in rootfs"));
+        }
+
+        info!(sl(), "use guest pause image cid {:?}", cid);
+        let pause_bundle = Path::new(CONTAINER_BASE).join(cid).join(target_subpath);
+        let pause_rootfs = pause_bundle.join("rootfs");
+        fs::create_dir_all(&pause_rootfs)?;
+
+        let copy_if_not_exists = |src: &Path, dst: &Path| -> Result<()> {
+            if !dst.exists() {
+                fs::copy(src, dst)?;
+            }
+            Ok(())
+        };
+        copy_if_not_exists(
+            &cc_pause_bundle.join(CONFIG_JSON),
+            &pause_bundle.join(CONFIG_JSON),
+        )?;
+        copy_if_not_exists(
+            &cc_pause_bundle.join("rootfs/pause"),
+            &pause_rootfs.join("pause"),
+        )?;
+
+        Ok(pause_rootfs.display().to_string())
+    }
+
+    /// pull_image is used for call image-rs to pull image in the guest.
+    /// # Parameters
+    /// - `image`: Image name (exp: quay.io/prometheus/busybox:latest)
+    /// - `cid`: Container id
+    /// - `image_metadata`: Annotations about the image (exp: "containerd.io/snapshot/cri.layer-digest": "sha256:24fb2886d6f6c5d16481dd7608b47e78a8e92a13d6e64d87d57cb16d5f766d63")
+    /// # Returns
+    /// - The image rootfs bundle path. (exp. /run/kata-containers/cb0b47276ea66ee9f44cc53afa94d7980b57a52c3f306f68cb034e58d9fbd3c6/images/rootfs)
+    pub async fn pull_image(
+        &self,
+        image: &str,
+        cid: &str,
+        image_metadata: &HashMap<String, String>,
+    ) -> Result<String> {
+        info!(sl(), "image metadata: {image_metadata:?}");
+        Self::set_proxy_env_vars();
```

Comment:
Does this function need to be executed everytime the `pull_image` is called? I mean the whole kata-agent process will share common `env`s. We only need to `set_proxy_env_vars` once in some place of the kata-gent code before pulling image.

Reply:
Fixed.

Reply:
hi @ChengyuZhu6 , could you please push the branch with this fix? `set_proxy_env_vars()` is still executed for every `pull_image` call.

Reply:
Hey @wainersm . In the code, we would only set the proxy once if the environment variable was detected.